### PR TITLE
KAA-524:Fixes notification duplication.

### PR DESCRIPTION
**Pull request by ohamov**

### CODE REVIEW

**Comment by ashvayka** [comment]
Please add same logic to C++ and Java clients

File: client/client-multi/client-c/src/kaa/collections/kaa_list.c [hunk 1]
```diff
@@ -263,3 +261,68 @@ void kaa_list_for_each(kaa_list_node_t *first, kaa_list_node_t *last, process_da
         first = kaa_list_next(first);
     }
 }
+
+static kaa_list_node_t *kaa_split_util(struct kaa_list_node_t *head);
+
+
+static kaa_list_node_t *kaa_merge_util(struct kaa_list_node_t *first, struct kaa_list_node_t *second, match_predicate pred)
+{
+    if (!first) {
```
Comment (ghost) [inline client/client-multi/client-c/src/kaa/collections/kaa_list.c]: Please, use KAA_RETURN_IF_NIL macros family.

File: client/client-multi/client-c/src/kaa/collections/kaa_list.c [hunk 2]
```diff
@@ -263,3 +261,68 @@ void kaa_list_for_each(kaa_list_node_t *first, kaa_list_node_t *last, process_da
         first = kaa_list_next(first);
     }
 }
+
+static kaa_list_node_t *kaa_split_util(struct kaa_list_node_t *head);
```
Comment (ghost) [inline client/client-multi/client-c/src/kaa/collections/kaa_list.c]: The 'struct' key word isn't needed.

File: client/client-multi/client-c/src/kaa/kaa_notification_manager.c [hunk 3]
```diff
@@ -99,6 +100,130 @@ static void shift_and_sub_extension(kaa_platform_message_reader_t *reader, uint3
...
+{
+    kaa_topic_notifications_node_t *node = (kaa_topic_notifications_node_t *)data;
+    return node->topic_id ==  *((uint64_t *) context);
+}
+
+static void kaa_destroy_notification_wrapper(void *data)
+{
+    KAA_RETURN_IF_NIL(data, );
+    kaa_notification_wrapper_t* wrapper = (kaa_notification_wrapper_t *)data;
+    if(wrapper->notification) {
```
Comment (ghost) [inline client/client-multi/client-c/src/kaa/kaa_notification_manager.c]: Add space.

File: client/client-multi/client-c/src/kaa/kaa_notification_manager.c [hunk 4]
```diff
@@ -99,6 +100,130 @@ static void shift_and_sub_extension(kaa_platform_message_reader_t *reader, uint3
...
+{
+    KAA_RETURN_IF_NIL(node, KAA_ERR_BADPARAM);
+    kaa_topic_notifications_node_t * new_node = (kaa_topic_notifications_node_t *) KAA_MALLOC(sizeof(kaa_topic_notifications_node_t));
+    if (!new_node) {
+        return KAA_ERR_NOMEM;
+    }
+
+    new_node->notifications = kaa_list_create();
+    if (!new_node->notifications) {
+        KAA_FREE(new_node);
```
Comment (ghost) [inline client/client-multi/client-c/src/kaa/kaa_notification_manager.c]: You may use kaa_destroy_notification_node() in all 'if' blocks.

File: client/client-multi/client-c/src/kaa/kaa_notification_manager.c [hunk 5]
```diff
@@ -1020,6 +1176,16 @@ kaa_error_t kaa_notification_manager_handle_server_sync(kaa_notification_manager
             }
         }
     }
+    if (kaa_list_get_size(self->notifications)) {
+        kaa_sort_notifications(self->notifications);
+        kaa_list_for_each(kaa_list_begin(self->notifications), kaa_list_back(self->notifications), &kaa_notify_notification_listeners, self);
+        kaa_list_destroy(self->notifications, &kaa_destroy_notification_node);
```
Comment (ghost) [inline client/client-multi/client-c/src/kaa/kaa_notification_manager.c]: Please, use kaa_list_clear() instead.

File: server/common/server-shared/src/main/java/org/kaaproject/kaa/server/sync/platform/BinaryEncDec.java [hunk 6]
```diff
@@ -437,7 +437,7 @@ private void encode(GrowingByteBuffer buf, NotificationServerSync notificationSy
             buf.put(NOTHING);
             buf.putShort((short) notificationSync.getNotifications().size());
             for (Notification nf : notificationSync.getNotifications()) {
-                buf.putInt(nf.getSeqNumber());
+                buf.putInt((nf.getSeqNumber() != null)? nf.getSeqNumber():0);
```
Comment (ghost) [inline server/common/server-shared/src/main/java/org/kaaproject/kaa/server/sync/platform/BinaryEncDec.java]: Add spaces.

File: client/client-multi/client-c/src/kaa/kaa_notification_manager.c [hunk 7]
```diff
@@ -99,6 +105,132 @@ static void shift_and_sub_extension(kaa_platform_message_reader_t *reader, uint3
...
+    kaa_topic_notifications_node_t * node = (kaa_topic_notifications_node_t *)data;
+    kaa_list_destroy(node->notifications, kaa_destroy_notification_wrapper);
+    KAA_FREE(node);
+}
+
+static kaa_error_t kaa_create_topic_notification_node(kaa_topic_notifications_node_t **node, kaa_notification_t *item, uint32_t *sqn, uint64_t *topic_id)
+{
+    KAA_RETURN_IF_NIL(node, KAA_ERR_BADPARAM);
+    kaa_topic_notifications_node_t * new_node = (kaa_topic_notifications_node_t *) KAA_MALLOC(sizeof(kaa_topic_notifications_node_t));
+    if (!new_node) {
```
Comment (ghost) [inline client/client-multi/client-c/src/kaa/kaa_notification_manager.c]: Replace with KAA_RETURN..

File: client/client-multi/client-c/src/kaa/kaa_notification_manager.c [hunk 8]
```diff
@@ -99,6 +105,132 @@ static void shift_and_sub_extension(kaa_platform_message_reader_t *reader, uint3
...
+    new_node->notifications = kaa_list_create();
+    if (!new_node->notifications) {
+        kaa_destroy_notification_node(new_node);
+        return KAA_ERR_NOMEM;
+    }
+
+    kaa_notification_wrapper_t * wrapper = (kaa_notification_wrapper_t *) KAA_MALLOC(sizeof(kaa_notification_wrapper_t));
+    if (!wrapper) {
+        kaa_list_destroy(new_node->notifications, NULL);
+        kaa_destroy_notification_node(new_node);
```
Comment (ghost) [inline client/client-multi/client-c/src/kaa/kaa_notification_manager.c:159]: Double notification list clear.

File: client/client-multi/client-c/src/kaa/kaa_notification_manager.c [hunk 9]
```diff
@@ -99,6 +105,132 @@ static void shift_and_sub_extension(kaa_platform_message_reader_t *reader, uint3
...
+        return KAA_ERR_NOMEM;
+    }
+
+    if(!kaa_list_push_back(new_node->notifications, wrapper)) {
+        kaa_list_destroy(new_node->notifications, NULL);
+        kaa_destroy_notification_node(new_node);
```
Comment (ghost) [inline client/client-multi/client-c/src/kaa/kaa_notification_manager.c:164]: The same as above.

File: client/client-multi/client-c/src/kaa/kaa_notification_manager.c [hunk 10]
```diff
@@ -99,6 +105,132 @@ static void shift_and_sub_extension(kaa_platform_message_reader_t *reader, uint3
...
+           new_wrapper->notification = item;
+           new_wrapper->sqn = *sqn;
+  }
+
+    return KAA_ERR_NONE;
+}
+
+static bool kaa_predicate_for_notifications(void *notif_1, void *notif_2)
+{
+    kaa_notification_wrapper_t *wrapper_1 = (kaa_notification_wrapper_t *)notif_1;
```
Comment (ghost) [inline client/client-multi/client-c/src/kaa/kaa_notification_manager.c:219]: Check on NULL.

File: client/client-multi/client-c/src/kaa/kaa_notification_manager.c [hunk 11]
```diff
@@ -867,10 +1010,41 @@ static kaa_error_t update_sequence_number(kaa_notification_manager_t *self, uint
...
 
+static void kaa_notify_notification_listeners(void *data, void* context)
+{
+   KAA_RETURN_IF_NIL2(data, context, );
+   kaa_error_t err = KAA_ERR_NONE;
+   kaa_topic_notifications_node_t *node = (kaa_topic_notifications_node_t *) data;
+   kaa_list_node_t *notification_list_node = kaa_list_begin(node->notifications);
+   kaa_notification_manager_t *self = (kaa_notification_manager_t *)context;
+   while(notification_list_node) {
+       KAA_LOG_TRACE(self->logger, KAA_ERR_NONE, "Passing notification\n");
```
Comment (ghost) [inline client/client-multi/client-c/src/kaa/kaa_notification_manager.c]: Please, remove this log trace.

**Pull request merged**